### PR TITLE
SAI versioning

### PR DESCRIPTION
**Pull request by ravi861**
Proposal and header file

### CODE REVIEW

File: saiversion.h [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+ */
+
+#if !defined (__SAIVERSION_H_)
+#define __SAIVERSION_H_
+
+#define SAI_MAJOR 1
+#define SAI_MINOR 7
+#define SAI_REVISION 1
+
+#define SAI_VERSION(maj, min, rev) (100 * (maj) + 10 * (min) + (rev))
```
Comment (lguohan) [inline saiversion.h:32]: for future extension, i think probably better to do 10000 * (major) + 100 * (min) + (rev) so the format is like

010701

File: doc/SAI-Proposal-versioning.md [hunk 2]
```diff
@@ -0,0 +1,70 @@
+SAI Versioning
+==============
+
```
Comment (lguohan) [inline doc/SAI-Proposal-versioning.md:3]: better to have authorship here.

**Comment by kcudnik** [comment]
just to point some of my thoughts on versioning: https://github.com/ravi861/SAI/pull/1#issuecomment-727950214

File: saiversion.h (hunk 1 above)
Comment (lguohan) [inline saiversion.h:32]: -> major, minor, revision

File: doc/SAI-Proposal-versioning.md [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+### Source Code
+A vendor publishes their SDK in source form and advice their users of which version of their SDK can be used for the requested SAI version. It is assumed the most common form this takes is each vendor maintains a trunk branch with the latest code and patches are back ported to specific branches created for each SAI version.
+
+> ### Example
+> Customer using Sonic version 201911 which is compatible with SAI 1.5.4. The vendor SDK version is 1.1. They customer wants to upgrade to a newer version of the SDK which supports SAI objects/attributes that are only supported in SDK 1.2. The common method to support this is to backport the supported attributes to 1.1.
+
+Proposal
+--------
+This proposal provides a static version file with version numbers hard-coded per release.
```
Comment (lguohan) [inline doc/SAI-Proposal-versioning.md:24]: can you share the the cmakefile or makefile you are referring.

File: saiversion.h [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+
+#define SAI_API_VERSION SAI_VERSION(SAI_MAJOR, SAI_MINOR, SAI_REVISION)
```
Comment (lguohan) [inline saiversion.h:34]: can we get a version comparing macro for user to use?

File: doc/SAI-Proposal-versioning.md (hunk 2 above)
Comment (ravi861) [reply doc/SAI-Proposal-versioning.md:13]: Added author.

File: saiversion.h (hunk 1 above)
Comment (ravi861) [reply saiversion.h]: Fixed

Comment (ravi861) [reply saiversion.h]: Done

File: doc/SAI-Proposal-versioning.md (hunk 3 above)
Comment (ravi861) [reply doc/SAI-Proposal-versioning.md:34]: Added an example for makefile. More can easily found for other build systems.

File: saiversion.h (hunk 4 above)
Comment (ravi861) [reply saiversion.h:35]: In my opinion, it is simple enough to use > < == for comparison in code as described in the doc. I can add more if required.
Did you have an API name in mind? SAI_API_VERSION_GREATER/LESSER?

File: saiversion.h [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+
+#if !defined (__SAIVERSION_H_)
+#define __SAIVERSION_H_
+
+#define SAI_MAJOR 1
+#define SAI_MINOR 7
+#define SAI_REVISION 1
+
+#define SAI_VERSION(major, minor, revision)                                    \
+  (10000 * (major) + 10 * (minor) + (revision))
```
Comment (kcudnik) [inline saiversion.h:33]: make this single line? also should this be:  (10000 * (major) + 100 * (minor) + (revision)) ??

Comment (ravi861) [reply saiversion.h]: I assumed there is an 80 char line limit. 1 line now. The version for v1.7.1 would be 10701 after making it 100.

File: doc/SAI-Proposal-versioning.md [hunk 6]
```diff
@@ -0,0 +1,98 @@
...
+
+### version.h
+```c
+#define SAI_MAJOR 1
+#define SAI_MINOR 7
+#define SAI_REVISION 1
+
+#define SAI_VERSION(major, minor, micro) \
+    (100 * (major) + 10 * (minor) + (micro))
```
Comment (lguohan) [inline doc/SAI-Proposal-versioning.md:43]: 10000 * (major) + 100 * (minor) + (micro)

**Review by kcudnik — APPROVED** [review]

**Comment by ravi861** [comment]
@lguohan @kcudnik can we please merge this or are there more comments?

**Comment by ravi861** [comment]
@lguohan Please let me know if you are waiting on something for merging this.

**Comment by kcudnik** [comment]
just noticed, why this file is in main dir? it should be in inc/

**Comment by ravi861** [comment]
> just noticed, why this file is in main dir? it should be in inc/

I didn't really have a preference. I kept it here since the file isn't really a SAI module and inc mostly includes those. Would you like it to move into inc?

**Comment by kcudnik** [comment]
i think it would make sense to have that file in inc/

**Pull request merged**